core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala

```diff
@@ -48,14 +50,25 @@ private[spark] class BlockStoreShuffleReader[K, C](
       // Note: we use getSizeAsMb when no suffix is provided for backwards compatibility
       SparkEnv.get.conf.getSizeAsMb("spark.reducer.maxSizeInFlight", "48m") * 1024 * 1024)
 
-    // Wrap the streams for compression based on configuration
-    val wrappedStreams = blockFetcherItr.map { case (blockId, inputStream) =>
-      blockManager.wrapForCompression(blockId, inputStream)
-    }
-
     val ser = Serializer.getSerializer(dep.serializer)
     val serializerInstance = ser.newInstance()
 
+    // Wrap the streams for compression and encryption based on configuration
+    val wrappedStreams = blockFetcherItr.map { case (blockId, inputStream) =>
+      val sparkConf = blockManager.conf
+      if (CryptoConf.isShuffleEncrypted(sparkConf)) {
+        if (SparkHadoopUtil.get.isYarnMode()) {
+          val cis = CryptoStreamUtils.createCryptoInputStream(inputStream, sparkConf)
+          blockManager.wrapForCompression(blockId, cis)
```

Comment:
Wouldn't it be more effective if you encrypted the compressed stream, instead of the other way around?